tests/checkpoint/test_fsdp_ckpt.py

```diff
@@ -0,0 +1,125 @@
+    mixed_precision = MixedPrecision(param_dtype=torch.bfloat16, reduce_dtype=torch.float32, buffer_dtype=torch.float32)
+
+    model = FSDP(model,
+                 use_orig_params=False,
+                 device_id=torch.cuda.current_device(),
+                 sharding_strategy=ShardingStrategy.FULL_SHARD,
+                 mixed_precision=mixed_precision,
+                 device_mesh=device_mesh)
+
+    optimizer = torch.optim.AdamW(model.parameters(), lr=1e-4)
+    lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1, gamma=0.9)
+
+    # Create checkpoint manager
+    tokenizer = AutoTokenizer.from_pretrained(model_name)
+    checkpoint_manager = FSDPCheckpointManager(model=model,
+                                               optimizer=optimizer,
+                                               lr_scheduler=lr_scheduler,
+                                               tokenizer=tokenizer)
+
+    # Generate sample input
+    batch_size = 2
+    seq_len = 32
+    vocab_size = 32000
+    # First input for initial update
+    input_ids1 = torch.randint(0, vocab_size, (batch_size, seq_len), device='cuda')
+    attention_mask1 = torch.ones_like(input_ids1)
+
+    # Second input for verification
+    input_ids2 = torch.randint(0, vocab_size, (batch_size, seq_len), device='cuda')
+    attention_mask2 = torch.ones_like(input_ids2)
+
+    # Step 1: Initial update and save checkpoint
+    outputs1 = model(input_ids=input_ids1, attention_mask=attention_mask1)
+    loss1 = outputs1.logits.mean()
+    loss1.backward()
+    optimizer.step()
+    lr_scheduler.step()
+    optimizer.zero_grad()
+
+    # Save checkpoint after first update
+    temp_dir = tempfile.mkdtemp()
+    checkpoint_path = os.path.join(temp_dir, 'checkpoint')
+    checkpoint_manager.save_checkpoint(local_path=checkpoint_path, hdfs_path=None, global_step=0)
+
+    # Step 2: Second update and forward pass
+    outputs2 = model(input_ids=input_ids2, attention_mask=attention_mask2)
+    loss2 = outputs2.logits.mean()
+    loss2.backward()
+    optimizer.step()
+    lr_scheduler.step()
+    optimizer.zero_grad()
+
+    # Record logits after second update
+    with torch.no_grad():
+        logits_before_load = model(input_ids=input_ids2, attention_mask=attention_mask2).logits
+
+    # Step 3: Load checkpoint and repeat second update
+    checkpoint_manager.load_checkpoint(checkpoint_path)
+
+    # Repeat the second update with same input
+    outputs3 = model(input_ids=input_ids2, attention_mask=attention_mask2)
+    loss3 = outputs3.logits.mean()
+    loss3.backward()
+    optimizer.step()
+    lr_scheduler.step()
+    optimizer.zero_grad()
+
+    # Record logits after loaded checkpoint and update
+    with torch.no_grad():
+        logits_after_load = model(input_ids=input_ids2, attention_mask=attention_mask2).logits
+
+    # Step 4: Verify outputs match
+    if rank == 0:
+        torch.testing.assert_close(logits_before_load, logits_after_load, atol=1e-2, rtol=1e-4)
```

Comment:
This should be compared for every rank? Also, it should be bitwise aligned if we toggle deterministic?

Reply:
bitwise align means using default atol and rtol of `torch.testing.assert_close`?

Reply:
Comparing with logits (not loss) should meet bitwise align requirement?